### ISSUE TITLE
Artifact generation process in the benchmark suite

### ISSUE THREAD

**Issue by pzread** [openxla/openxla-benchmark#12]
Models usually require the raw inputs (text, images) to be preprocessed (e.g., tokenizer).

Currently we have `ModelTestDataArtifact` to store the preprocessed data for each model, but we should also include the raw data and the parameters to reproduce the preprocessed data from the raw data.

The same process also needs to handle the exported model generation.

**Comment by pzread** [comment]
For multi-tensor inputs, we probably also want to pack them in a single archive instead of multiple `.npy` in GCS. Since there is probably no reason to download them separately and makes artifact management simpler.

**Comment by pzread** [comment]
Some notes from offline discussions:

Since a model implemented with frameworks usually has parameters to configure input seqlen, batch size, etc., we abstract it with two classes: `ModelImplementation` and `Model`. The `ModelImplementation` describes the configurable model implementation and `Model` describes a `ModelImplementation` with concrete model parameters to initialize it.

```
ModelImplementation: X

Model: X_512XI32_BATCH1
- model_impl: X
- model_parameters: {seqlen: 512, batch: 1}

Model: X_1024XI32_BATCH16
- model_impl: X
- model_parameters: {seqlen: 1024, batch: 16}
```

Therefore, to generate a exported model (e.g. stablehlo), the process requires to first initialize the model with the specific model parameters (so the input seqlen, batch size are determined), then export the initialized model.

The preprocessed input data follows the same process. Since the raw input data needs to be preprocessed corresponding to the model (e.g. if the model seqlen is 512, the input data needs to be padding to 512), each preprocessed input data is **model-specific**. This also applies to expected outputs.

Therefore, `ModelTestData` should contain two fields: `raw_data_url` to download the raw source data and `model` to know which `Model` is used to preprocess the data. As a result, the structure looks like this:

```
ModelImplementation: X

Model: X_512XI32_BATCH1
- model_impl: X
- model_parameters: {seqlen: 512, batch: 1}

ModelTestData: INPUT_DATA_X_512XI32_BATCH1_SIMPLE_TEXT
- model: X_512XI32_BATCH1
- raw_data_url: gs://.../simple_text.txt

Model: X_1024XI32_BATCH16
- model_impl: X
- model_parameters: {seqlen: 1024, batch: 16}

ModelTestData: INPUT_DATA_X_1024XI32_BATCH16_SIMPLE_TEXT
- model: X_1024XI32_BATCH16
- raw_data_url: gs://.../simple_text.txt
```

We store artifacts on GCS. In order to version the exported models/preprocessed data, here is the proposal:

`${framework}/${10 digits of git hash}_${timestamp}` as prefix in the path, followed by `${model_name}` and `${model_name}/${test_data_name}`. Here are the reasons:

- `${framework}/` to categorize artifacts based on the ML framework
- `${10 digits of git hash}` to identify code version
- `${timestamp}` to help generate multiple times with the same code version

For example:

```
gs://.../jax/6d8d59e1c6_1688074165/X_512XI32_BATCH1/stablehlo.mlir
gs://.../jax/6d8d59e1c6_1688074165/X_512XI32_BATCH1/INPUT_DATA_X_1024XI32_BATCH16_SIMPLE_TEXT/input_npy.tgz
gs://.../jax/6d8d59e1c6_1688074165/X_512XI32_BATCH1/OUTPUT_DATA_X_1024XI32_BATCH16_SIMPLE_TEXT/output_npy.tgz
```

**Comment by pzread** [comment]
@mariecwhite let me know what you think :slightly_smiling_face: 

Also @reichlfl who might have opinions on this topic

**Comment by mariecwhite** [comment]
+1 to the suggestions.

I think there is a level of abstraction missing between downloading the raw data URLs and calling the model's `preprocess()` method. We need some kind of DataProvider class that knows how to read in the raw data and call `preprocess()` with the parameters that it expects.

Some examples:

1. `TextDataProvider` downloads a .txt file and loads it into a string.
2. `ImageDataProvider` downloads a .png, .jpeg, etc, loads and resizes the image into the expected width and height. Possibly changing the data layout e.g. yuv, rgb, etc.
3. `ImageTrainingDataProvider` would provide batched image examples used in training. A concrete example is the  [ImageNet](https://www.kaggle.com/c/imagenet-object-localization-challenge/overview/description) dataset which is distributed in the form of a  directory of jpeg images. `ImageTrainingDataProvider` could load a batch of these images in a format that's expected by the downstream `preprocess()` function.

In terms of freezing the python package versions, maybe we offer two setup_venv.sh modes (or separate files): one frozen and one bleeding edge version.

**Comment by pzread** [comment]
@mariecwhite Anything left to do on this issue?

**Comment by mariecwhite** [comment]
We are good here. We've regenerated the PyTorch, TF, JAX artifacts from the new repo and updated the links.
